src/phoenix/db/insertion/trace_annotation.py

```diff
@@ -0,0 +1,139 @@
+        List[Received[Insertables.TraceAnnotation]],
+        List[Postponed[Precursors.TraceAnnotation]],
+        List[Received[Precursors.TraceAnnotation]],
+    ]:
+        to_insert: List[Received[Insertables.TraceAnnotation]] = []
+        to_postpone: List[Postponed[Precursors.TraceAnnotation]] = []
+        to_discard: List[Received[Precursors.TraceAnnotation]] = []
+
+        stmt = _select_existing(*map(_key, parcels))
+        existing = [_ async for _ in await session.stream(stmt)]
+        existing_traces: Mapping[str, _TraceAttr] = {
+            trace_id: _TraceAttr(trace_rowid) for trace_rowid, trace_id, *_ in existing
+        }
+        existing_annos: Mapping[_Key, _AnnoAttr] = {
+            (name, trace_id): _AnnoAttr(trace_rowid, id_, updated_at)
+            for trace_rowid, trace_id, id_, name, updated_at in existing
+            if id_ is not None
+        }
+
+        for p in parcels:
+            if (anno := existing_annos.get(_key(p))) is not None:
+                if p.received_at <= anno.updated_at:
+                    to_discard.append(p)
+                else:
+                    to_insert.append(
+                        Received(
+                            received_at=p.received_at,
+                            item=p.item.as_insertable(
+                                trace_rowid=anno.trace_rowid,
+                                id_=anno.id_,
+                            ),
+                        )
+                    )
+            elif (trace := existing_traces.get(p.item.trace_id)) is not None:
+                to_insert.append(
+                    Received(
+                        received_at=p.received_at,
+                        item=p.item.as_insertable(
+                            trace_rowid=trace.trace_rowid,
+                        ),
+                    )
+                )
+            elif isinstance(p, Postponed):
+                if p.retries_left > 1:
+                    to_postpone.append(p.postpone(p.retries_left - 1))
+                else:
+                    to_discard.append(p)
+            elif isinstance(p, Received):
+                to_postpone.append(p.postpone(retry_allowance))
+            else:
+                to_discard.append(p)
+
+        assert len(to_insert) + len(to_postpone) + len(to_discard) == len(parcels)
+        to_insert = dedup(sorted(to_insert, key=_time, reverse=True), _unique_by)[::-1]
+        return to_insert, to_postpone, to_discard
+
+
+def _select_existing(
+    *identifiers: Tuple[str, str],
+) -> Select[Tuple[int, str, int, str, datetime]]:
+    existing_traces = (
+        select(models.Trace.id, models.Trace.trace_id)
+        .where(models.Trace.trace_id.in_({trace_id for _, trace_id in identifiers}))
+        .cte()
+    )
+    table = models.TraceAnnotation
+    return select(
+        existing_traces.c.id,
+        existing_traces.c.trace_id,
+        table.id,
+        table.name,
+        table.updated_at,
+    ).outerjoin_from(
```

Comment:
This strikes me as a good place to use `with_only_columns` to simplify the return type of the function. Same for trace and doc annotations.

Reply:
after reading the docs, i'm still not clear how it can be used